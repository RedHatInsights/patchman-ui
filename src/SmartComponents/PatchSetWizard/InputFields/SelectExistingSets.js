import React, { useState, useMemo, useEffect } from 'react';
import propTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Select, SelectOption, SelectVariant, FormGroup, Spinner } from '@patternfly/react-core';
import useFormApi from '@data-driven-forms/react-form-renderer/use-form-api';
import { fetchPatchSetsAction, changePatchSetsParams, clearPatchSetsAction } from '../../../store/Actions/Actions';

const SelectExistingSets = ({ setSelectedPatchSet, selectedSets, systems }) => {
    const dispatch = useDispatch();
    const formOptions = useFormApi();
    const [isOpen, setOpen] = useState(true);

    const rows = useSelector(({ PatchSetsStore }) => PatchSetsStore.rows);
    const queryParams = useSelector(({ PatchSetsStore }) => PatchSetsStore.queryParams);
    const status = useSelector(({ PatchSetsStore }) => PatchSetsStore.status);

    useEffect(() => {
        dispatch(fetchPatchSetsAction({ ...queryParams, limit: 10 }));

        return () => dispatch(clearPatchSetsAction());
    }, [queryParams.page]);

    const patchOptions = useMemo(() =>{
        if (status.isLoading) {
            return [<SelectOption key='loading'><Spinner size="md"/></SelectOption>];
        }

        return rows.map(set => <SelectOption key={set.id} value={set.name} />);
    }, [rows, status.isLoading]);

    const handleOpen = () => {
        setOpen(!isOpen);
    };

    const handleSelect = (_, selected) =>{
        setOpen(false);
        setSelectedPatchSet(selected);

        const selectedSet = rows.filter(set => set.name === selected);
        if (selectedSet.length === 1) {
            formOptions.change('existing_patch_set', { name: selectedSet[0]?.name, systems, id: selectedSet[0]?.id });
        }

    };

    const onViewMoreClick = () => {
        dispatch(changePatchSetsParams({ ...queryParams, page: queryParams.page + 1 }));
    };

    return (
        <FormGroup fieldId="existing_patch_set" label="Choose a Patch set" isRequired>
            <Select
                variant={SelectVariant.single}
                aria-label="Select Input"
                onSelect={handleSelect}
                placeholderText="Select an option"
                selections={selectedSets}
                onToggle={handleOpen}
                isOpen={isOpen}
                isDisabled={false}
                loadingVariant={{
                    text: 'View more',
                    onClick: onViewMoreClick
                }}
            >
                {patchOptions}
            </Select>
        </FormGroup >
    );
};

SelectExistingSets.propTypes = {
    setSelectedPatchSet: propTypes.func,
    selectedSets: propTypes.array,
    systems: propTypes.array
};
export default SelectExistingSets;
