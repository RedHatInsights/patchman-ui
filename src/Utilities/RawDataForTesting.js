// raw data for testing purposes only

export const advisoryRows = [{
    attributes: {
        advisory_type: 1,
        applicable_systems: 1,
        description: 'The microcode_ctl ',
        public_date: '2020-06-24T17:22:25Z',
        synopsis: 'microcode_ctl',
        severity: 3
    },
    id: 'RHEA-2020:2743',
    type: 'advisory'
}];

export const systemAdvisoryRows = [{
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
}];

export const systemRows = [{
    attributes: {
        display_name: 'automation_host',
        last_evaluation: '2020-08-12T13:57:54.028883Z',
        last_upload: '2020-08-12T09:26:33.891907Z',
        packages_installed: 0,
        packages_updatable: 0,
        rhba_count: 213,
        rhea_count: 35,
        rhsa_count: 77,
        stale: false
    },
    id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
    type: 'system'
}];

export const advisoryHeader = {
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        modified_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
};

export const advisoryDetailRows = {
    data: {
        attributes: {
            cves: ['CVE-2020-11080'],
            description: 'libnghttp2.',
            fixes: null,
            modified_date: '2020-06-25T16:31:29Z',
            packages: { libnghttp2: '1.33.0-3.el8_2.1.x86_64' },
            public_date: '2020-06-25T16:31:29Z',
            references: [],
            severity: 3,
            solution: 'For details',
            synopsis: 'Important: ',
            topic: 'An update .'
        }
    },
    id: 'RHSA-2020:2755',
    type: 'advisory'
};
export const systemPackages = [{
    description: 'This package contains testing data',
    evra: '2.2.*',
    name: 'acl',
    summary: 'Access control list utilities',
    updates: null
}];

export const entityDetail = {
    loaded: true,
    activeApps: [],
    entity: {
        account: '60'
    }
};
