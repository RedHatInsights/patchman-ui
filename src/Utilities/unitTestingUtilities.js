import { useSelector } from 'react-redux';

/* eslint-disable */
export const initMocks = () => {

    window.insights = {
        loadInventory: (args) => new Promise((resolve, reject) =>  {
            resolve(({
                inventoryConnector: (args) => {
                            const InventoryTable = ({ children }) => <div >A mock passed! {children} </div>;
                            return ({ InventoryTable });
                        },
                        mergeWithEntities: () => {}
            }));
        }),
        loadRemediations: () => new Promise((resolve, reject) =>  {
            resolve({ 
                openWizard: () => new Promise((resolve, reject) =>  { 
                    resolve(true)
                })
            });
        })
    };

};

export const mockStore = (initialState, mutatedState) => {
    const customMiddleWare = store => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mutatedState);
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);

    return mockStore(initialState);
}
/* eslint-enable */
