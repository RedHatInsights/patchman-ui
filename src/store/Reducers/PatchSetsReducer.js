import * as ActionTypes from '../ActionTypes';
import {
    changeFilters,
    selectRows,
    fetchPending,
    fetchRejected
} from './HelperReducers';

// Initial State. It should not include page and perPage to persist them dynamically
export const initialState = {
    rows: [],
    selectedRows: [],
    queryParams: {
        page: 1,
        perPage: 20,
        filter: {}
    },
    status: {},
    metadata: {
        limit: 25,
        offset: 0,
        total_items: 0
    },
    error: {}
};

export const PatchSetsReducer = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.FETCH_ALL_PATCH_SETS + '_FULFILLED':
            return {
                ...state,
                rows: action.payload.data.map(set => ({ ...set.attributes, id: set.id })),
                metadata: action.payload.meta || {},
                error: {},
                status: { code: action.payload.status, isLoading: false, hasError: false }
            };

        case ActionTypes.FETCH_ALL_PATCH_SETS + '_PENDING':
            return fetchPending(state);

        case ActionTypes.FETCH_ALL_PATCH_SETS + '_REJECTED':
            return fetchRejected(state, action);

        case ActionTypes.CHANGE_PATCH_SET_PARAMS:
            console.log('changepara', action);
            return changeFilters(state, action);

        case ActionTypes.SELECT_PATCH_SET_ROW:
            return selectRows(state, action);

        default:
            return state;
    }
};
