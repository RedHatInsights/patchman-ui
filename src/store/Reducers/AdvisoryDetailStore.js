import { STATUS_LOADING, STATUS_RESOLVED } from '../../Utilities/constants';
import { CLEAR_ADVISORY_DETAILS, FETCH_ADVISORY_DETAILS } from '../ActionTypes';
import {

    fetchPending,
    fetchRejected
} from './HelperReducers';

export let initialState = {
    data: { attributes: {} },
    status: STATUS_LOADING
};

// Reducer
export const AdvisoryDetailStore = (state = initialState, action) => {
    let newState = { ...state };
    switch (action.type) {
        case FETCH_ADVISORY_DETAILS + '_FULFILLED':
            return {
                ...state,
                status: STATUS_RESOLVED,
                data: action.payload.data,
                error: {}
            };

        case FETCH_ADVISORY_DETAILS + '_PENDING':
            return fetchPending(newState);

        case FETCH_ADVISORY_DETAILS + '_REJECTED':
            return fetchRejected(newState, action);

        case CLEAR_ADVISORY_DETAILS:
            return initialState;

        default:
            return state;
    }
};
