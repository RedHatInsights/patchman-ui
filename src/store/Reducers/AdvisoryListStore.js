import { storeListDefaults } from '../../Utilities/constants';
import * as ActionTypes from '../ActionTypes';
import {
    changeFilters,
    selectRows,
    expandRows,
    fetchPending,
    fetchRejected,
    fetchFulfilled
} from './HelperReducers';

export const AdvisoryListStore = (state = storeListDefaults, action) => {
    let newState = { ...state };
    switch (action.type) {
        case ActionTypes.FETCH_APPLICABLE_ADVISORIES + '_FULFILLED':
            return fetchFulfilled(newState, action);

        case ActionTypes.FETCH_APPLICABLE_ADVISORIES + '_PENDING':
            return fetchPending(newState);

        case ActionTypes.FETCH_APPLICABLE_ADVISORIES + '_REJECTED':
            return fetchRejected(newState, action);

        case ActionTypes.CHANGE_ADVISORY_LIST_PARAMS:
            return changeFilters(newState, action);

        case ActionTypes.EXPAND_ADVISORY_ROW:
            return expandRows(newState, action);

        case ActionTypes.SELECT_ADVISORY_ROW:
            return selectRows(newState, action);

        case ActionTypes.TRIGGER_GLOBAL_FILTER:
            return changeFilters(newState, action);

        default:
            return state;
    }
};
