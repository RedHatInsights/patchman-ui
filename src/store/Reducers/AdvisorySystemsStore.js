import * as ActionTypes from '../ActionTypes';
import { changeFilters } from './HelperReducers';
import { changeListParams } from '../../Utilities/Helpers';

// Initial State. It should not include page and perPage to persist them dynamically
const initialState = {
    selectedRows: {},
    queryParams: {
        page: 1,
        perPage: 20
    }
};
// Reducer

export const AdvisorySystemsStore = (state = initialState, action) => {
    let newState = { ...state };
    switch (action.type) {

        case ActionTypes.CHANGE_AFFECTED_SYSTEMS_PARAMS:
            delete action.payload?.filter;
            newState.queryParams = changeListParams(newState.queryParams, action.payload);
            return newState;

        case ActionTypes.TRIGGER_GLOBAL_FILTER:
            return changeFilters(newState, action);

        case ActionTypes.CLEAR_ADVISORY_SYSTEMS_REDUCER:
            return initialState;

        default:
            return state;
    }
};

