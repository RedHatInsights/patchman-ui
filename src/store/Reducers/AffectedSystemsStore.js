import { storeListDefaults } from '../../Utilities/constants';
import * as ActionTypes from '../ActionTypes';
import {
    changeFilters,
    selectRows,
    fetchPending,
    fetchRejected,
    fetchFulfilled
} from './HelperReducers';

export const AffectedSystemsStore = (state = storeListDefaults, action) => {
    let newState = { ...state };
    switch (action.type) {
        case ActionTypes.FETCH_AFFECTED_SYSTEMS + '_PENDING':
            return fetchPending(newState);

        case ActionTypes.FETCH_AFFECTED_SYSTEMS + '_REJECTED':
            return fetchRejected(newState, action);

        case 'SELECT_ENTITY':
            return selectRows(newState, action);

        case ActionTypes.CHANGE_AFFECTED_SYSTEMS_PARAMS:
            return changeFilters(newState, action);

        case ActionTypes.FETCH_AFFECTED_SYSTEMS + '_FULFILLED':
            return fetchFulfilled(newState, action);

        case ActionTypes.CLEAR_AFFECTED_SYSTEMS:
            return storeListDefaults;

        case ActionTypes.TRIGGER_GLOBAL_FILTER:
            return changeFilters(newState, action);

        default:
            return state;
    }
};
