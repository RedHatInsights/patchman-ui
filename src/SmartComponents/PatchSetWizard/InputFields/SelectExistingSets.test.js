import { Provider } from 'react-redux';
import { useSelector } from 'react-redux';
import configureStore from 'redux-mock-store';
import { act } from 'react-dom/test-utils';
import { BrowserRouter as Router } from 'react-router-dom';
import { Select, SelectOption } from '@patternfly/react-core';

import SelectExistingSets from './SelectExistingSets';
import { initialState } from '../../../store/Reducers/PatchSetsReducer';
import { initMocks } from '../../../Utilities/unitTestingUtilities.js';
import { patchSets } from '../../../Utilities/RawDataForTesting';

initMocks();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
}));

const mockChage = jest.fn((payload) => { return payload; });

jest.mock('@data-driven-forms/react-form-renderer/use-form-api', () => () => ({
    change: mockChage
})
);

const initStore = (state) => {
    const customMiddleWare = () => next => action => {
        useSelector.mockImplementation(callback => {
            return callback({ PatchSetsStore: state });
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);
    return mockStore({ PatchSetsStore: state });
};

let wrapper;
let store = initStore(initialState);

beforeEach(() => {
    store.clearActions();
    useSelector.mockImplementation(callback => {
        return callback({
            PatchSetsStore: {
                ...initialState,
                rows: patchSets,
                status: { ...initialState, isLoading: false }
            }
        });
    });
    wrapper = mount(<Provider store={store}>
        <Router>
            <SelectExistingSets
                setSelectedPatchSet={jest.fn()}
                systems={['test-system']}
            />
        </Router>
    </Provider>);
});

describe('SelectExistingSets', () => {
    it('Should fetch patch-sets on mount with limit 10', () =>{
        const actions = store.getActions();
        expect(actions[0].type).toEqual('FETCH_ALL_PATCH_SETS');
    });

    it('Should calculate Select options after patch-sets are loaded', () => {
        const selectOptions = wrapper.find(SelectOption);

        expect(selectOptions).toMatchSnapshot();
    });

    it('Should handle patch-set selection', () => {
        act(() => wrapper.find(Select).props().onSelect(null, 'test-set-1'));
        expect(mockChage).toHaveBeenCalledWith(
            'existing_patch_set',
            { id: 1, name: 'test-set-1', systems: ['test-system'] }
        );
    });
});
