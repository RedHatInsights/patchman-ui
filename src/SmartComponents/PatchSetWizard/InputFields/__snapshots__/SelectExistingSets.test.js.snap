// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectExistingSets Should calculate Select options after patch-sets are loaded 1`] = `ReactWrapper {}`;
